code/controllers/subsystems/typing.dm

```diff
@@ -0,0 +1,208 @@
+SUBSYSTEM_DEF(typing)
+	name = "Typing"
+	flags = SS_BACKGROUND | SS_NO_INIT
+	wait = 0.5 SECONDS
+
+	/// The skin control to poll for TYPING_STATE_INPUT status.
+	var/const/INPUT_HANDLE = "outputwindow.input"
+
+	var/const/INFLIGHT_TIMEOUT = 5 SECONDS
+
+	/// The status entry index of the related client's typing indicator visibility preference.
+	var/const/INDEX_PREFERENCE = 1
+
+	/// The status entry index of the inflight state.
+	var/const/INDEX_INFLIGHT = 2
+
+	/// The status entry index of the timeout threshold.
+	var/const/INDEX_TIMEOUT = 3
+
+	/// The status entry index of the input bar typing state.
+	var/const/INDEX_INPUT_STATE = 4
+
+	/// The status entry index of the verb input typing state.
+	var/const/INDEX_VERB_STATE = 5
+
+	/// The highest index in a status entry.
+	var/const/MAX_INDEX = 5
+
+	/*
+	* A list of (ckey = list(
+			preference = 0|1,
+			inflight = 0|1,
+			timeout = num,
+			istyping_input = 0|1,
+			istyping_hotkey = 0|1
+		), ...)
+		See .proc/get_entry for details.
+	*/
+	var/static/list/status = list()
+
+	/// Matches input bar verbs that should set TYPING_STATE_INPUT.
+	var/static/regex/match_verbs = regex("^(Me|Say) +\"?\\w+")
+
+	/// A list of clients waiting to be polled for input state.
+	var/static/list/client/queue = list()
```

Comment:
Why is this (and status) static? It only affects Recover...but does so negatively, in that you need extra Recover handling because for some reason these are static.